src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs

```diff
@@ -749,4 +788,35 @@ public override IReadOnlyList<Type> GetParameterTypes(string methodName)
 
         return null;
     }
+
+    // Starts an Activity for a Hub method invocation and sets up all the tags and other state.
+    // Make sure to call Activity.Stop() once the Hub method completes, and consider calling SetActivityError on exception.
+    private static Activity? StartActivity(HubConnectionContext connectionContext, IServiceProvider serviceProvider, string methodName)
+    {
+        if (serviceProvider.GetService<SignalRActivitySource>() is SignalRActivitySource signalRActivitySource
```

Comment:
Can we do this once? Does this need to be per invocation? (We can do this change as a follow up).

Reply:
We don't have a service provider until this point. We could do the whole `_activitySource ??= serviceProvider.GetService<SignalRActivitySource>()` pattern to only do it once, but that relies on `SignalRActivitySource` always being a singleton (at least with MEDI, idk how other containers work). This is true today, but if we ever make it public that would be a concern.